### PR TITLE
Add remove nulls function to commonserver

### PR DESCRIPTION
**Pull request by ronykoz**
## Status
Ready

## Related Issues
fixes: https://github.com/demisto/etc/issues/16386

## Description
Add remove nulls function to commonserver

### CODE REVIEW

File: Scripts/script-CommonServerPython.yml [hunk 1]
```diff
@@ -773,6 +773,17 @@ script: |-
 
   tblToMd = tableToMarkdown
 
+  def removeNullsFromDictionary(dict):
```
Comment (anara123) [inline Scripts/script-CommonServerPython.yml]: removeNullsFromDictionary => remove_nulls_from_dictionary

File: Scripts/script-CommonServerPython.yml [hunk 2]
```diff
@@ -773,6 +773,17 @@ script: |-
...
+      """
+          Remove Null values from a dictionary. (updating the given dictionary)
+
+          :type data: ``dict`` or ``list``
```
Comment (yaakovi) [inline Scripts/script-CommonServerPython.yml]: you didn't handle the list option

**Comment by ronykoz** [comment]
please note that i removed the CompareDates function as it was not working, and required extra package which our base image did not have

File: Scripts/CommonServerPython/CommonServerPython.py [hunk 3]
```diff
@@ -0,0 +1,1580 @@
+import demistomock as demisto
+from CommonServerPython import *
```
Comment (ronykoz) [inline Scripts/CommonServerPython/CommonServerPython.py]: remove this line

File: Scripts/CommonServerPython/CommonServerPython_test.py [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+    # sanity
+    table = tableToMarkdown('tableToMarkdown test', data)
+    expected_table = '''### tableToMarkdown test
+|header_2|header_3|header_1|
+|---|---|---|
+|b1|c1|a1|
+|b2|c2|a2|
+|b3|c3|a3|
+'''
+    tables.append((table, expected_table, ))
```
Comment (ronykoz) [inline Scripts/CommonServerPython/CommonServerPython_test.py]: replace it with assert

File: Scripts/CommonServerPython/CommonServerPython_test.py [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+def test_append_context():
+    appendContext('empty_string_key', '')
+    appendContext('empty_list_key', [])
+    appendContext('zero_key', 0)
+    appendContext('none_key', None)
+    demisto.results("placeholder so the run would finish")
+
+
+class TestIsError(unittest.TestCase):
```
Comment (ronykoz) [inline Scripts/CommonServerPython/CommonServerPython_test.py]: replace with pytest

**Comment by ronykoz** [comment]
Tested on demisto/python:1.0 demisto/python3

File: .gitignore [hunk 6]
```diff
@@ -7,7 +7,7 @@ TestData/EmailWithNonUnicodeSubject.eml
 *.pyc
 .pytest_cache
 
-CommonServerPython.py
+./CommonServerPython.py
```
Comment (glicht) [inline .gitignore]: This is not good. As part of the way we run unit tests and linting we copy the common server file to the package dir. These files now appear in git status. I suggest we stay with the way it was. You can still commit the CommonServerPython file for the package with `--force`.

File: Tests/scripts/pkg_dev_tasks_in_docker.py [hunk 7]
```diff
@@ -115,11 +115,9 @@ def docker_run(project_dir, docker_image, no_test, no_lint, keep_container):
     pylint_files = get_pylint_files(project_dir)
     # copy demistomock and common server
     shutil.copy(CONTENT_DIR + '/Tests/demistomock/demistomock.py', project_dir)
-    open(project_dir + '/CommonServerUserPython.py', 'a').close()  # create empty file
     shutil.rmtree(project_dir + '/__pycache__', ignore_errors=True)
-    subprocess.check_call(['python2', CONTENT_DIR + '/package_extractor.py', '-i',
-                           'Scripts/script-CommonServerPython.yml', '-o',
-                           project_dir + '/CommonServerPython.py'], cwd=CONTENT_DIR)
+    subprocess.check_call(['cp', CONTENT_DIR + '/Scripts/CommonServerPython/CommonServerPython.py',
+                           CONTENT_DIR + '/CommonServerPython.py'], cwd=CONTENT_DIR)
```
Comment (glicht) [inline Tests/scripts/pkg_dev_tasks_in_docker.py]: We need the file in the project dir for the way we run within docker.
```suggestion
                           project_dir + '/CommonServerPython.py'], cwd=CONTENT_DIR)
```

File: .gitignore (hunk 6 above)
Comment (ronykoz) [reply .gitignore]: The copying shouldn't add them to the git status. I think this way is better because we will keep working regularly with CommonServerPython as with all the other files in the system

**Comment by ronykoz** [comment]
Made sure all the recent changes for the file are in the package, everything seem to work fine.
Only having problems with mypy which fails on local env but works fine on Circle(Consulting with @glicht about this)
Is it good to go?

**Comment by glicht** [comment]
I am going to commit a change that when commonserver is modified all unit tests are run. Once that is in and we see that all unit tests are good we can merge

**Comment by glicht** [comment]
@ronykoz @yaakovi  @anara123  see my commits. I also put in some changes in common server regrading the way we treat `bytes` in py3. This is as a result of running mypy with py3 on common server (we don't have this in the build, yet!).

**Comment by glicht** [comment]
@ronykoz why is the build failing on playbooks with uuids (such as 97393cfc-2fc4-4dfe-8b6e-af64067fc436)?

**Comment by yaakovi** [comment]
@glicht those are AWS tests, they already been moved to fix until the AWS PR is merged

File: package_creator.py [hunk 8]
```diff
@@ -130,9 +130,12 @@ def get_code_file(package_path, script_type):
     :rtype: str
     """
 
-    ignore_regex = r'^CommonServerPython\.py|^CommonServerUserPython\.py|demistomock\.py|test_.*\.py|_test\.py'
+    ignore_regex = r'CommonServerPython\.py|CommonServerUserPython\.py|demistomock\.py|test_.*\.py|_test\.py'    
     if not package_path.endswith('/'):
         package_path += '/'
+    if package_path.endswith('Scripts/CommonServerPython/'):
```
Comment (yaakovi) [inline package_creator.py]: shouldn't this be `startswith`?

File: Integrations/ANYRUN/ANYRUN.py [hunk 9]
```diff
@@ -71,7 +71,7 @@ def underscore_to_camel_case(s):
         The converted string (e.g. heLLoWorld).
     """
 
-    if not isinstance(s, STRING_TYPES):
+    if not isinstance(s, (str, bytes)):
```
Comment (yaakovi) [inline Integrations/ANYRUN/ANYRUN.py]: why did you change that?
the value of `STRING_TYPES` is set to that value incase of python3

File: Scripts/CommonServerPython/CommonServerPython.py [hunk 10]
```diff
@@ -0,0 +1,1586 @@
...
+import re
+from collections import OrderedDict
+
+import xml.etree.cElementTree as ET
+
+IS_PY3 = sys.version_info[0] == 3
+# pylint: disable=undefined-variable
+if IS_PY3:
+    STRING_TYPES = (str, bytes)  # type: ignore
+    STRING_OBJ_TYPES = (str)
```
Comment (yaakovi) [inline Scripts/CommonServerPython/CommonServerPython.py]: in order for the type be tuple (consistent with the else part)
```suggestion
    STRING_OBJ_TYPES = (str, )
```

File: package_creator.py (hunk 8 above)
Comment (glicht) [reply package_creator.py]: It's ok. We get a full path of the sort `/usr/.../Scripts/CommonServerPython`

File: Tests/scripts/configure_tests.py [hunk 11]
```diff
@@ -22,8 +22,9 @@
 
 # File names
 ALL_TESTS = ["scripts/script-CommonIntegration.yml", "scripts/script-CommonIntegrationPython.yml",
-             "scripts/script-CommonServer.yml", "scripts/script-CommonServerPython.yml",
-             "scripts/script-CommonServerUserPython.yml", "scripts/script-CommonUserServer.yml"]
+             "scripts/script-CommonServer.yml", "scripts/script-CommonServerPython_4_1.yml",
```
Comment (yaakovi) [inline Tests/scripts/configure_tests.py]: needs to update this list (remove the 4.1)

**Comment by ronykoz** [comment]
@glicht i think we can merge this thing

**Comment by glicht** [comment]
We will merge on Monday
